### PR TITLE
feat: CSM contracts as a separate module

### PR DESCRIPTION
**Pull request by madlabman**

### CODE REVIEW

File: src/providers/execution/contracts/CSModule.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            for e in cast(ContractEvent, self.events.StuckSigningKeysCountChanged).get_logs(
+                fromBlock=l_block, toBlock=r_block
+            ):
+                yield e
+
+            if not self.is_deployed(l_block):
+                break
+
+            r_block = l_block
+            l_block = l_block - variables.EVENTS_SEARCH_STEP
```
Comment (vgorkavenko) [inline src/providers/execution/contracts/CSModule.py:54]:
```suggestion
            l_block = r_block - variables.EVENTS_SEARCH_STEP
```
just to be aligned with 41 line

File: src/utils/blockstamp.py [hunk 2]
```diff
@@ -25,12 +25,15 @@ def build_blockstamp(
...
+
+    execution_payload = slot_details.message.body["execution_payload"]
 
     return {
-        'slot_number': SlotNumber(int(slot_details.message.slot)),
-        'state_root': slot_details.message.state_root,
-        'block_number': BlockNumber(int(execution_data['block_number'])),
-        'block_hash': execution_data['block_hash'],
-        'block_timestamp': Timestamp(int(execution_data['timestamp']))
+        "slot_number": SlotNumber(int(slot_details.message.slot)),
```
Comment (vgorkavenko) [inline src/utils/blockstamp.py:34]: why `"`? looks like we use `'` everywhere

File: src/providers/execution/contracts/CSModule.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            ):
+                continue
+
+            yield op_id
+
+    def get_stuck_keys_events(self, block: BlockIdentifier) -> Iterable[EventData]:
+        """Fetch all the StuckSigningKeysCountChanged events up to the given block (closed interval)"""
+
+        r_block = int(self.w3.eth.get_block(block)["number"])  # type: ignore
+        l_block = r_block - variables.EVENTS_SEARCH_STEP
```
[the remark this answers is not in the document]
Comment (madlabman) [reply src/providers/execution/contracts/CSModule.py:41]: No. And this one is a pretty generic one.

File: src/utils/blockstamp.py (hunk 2 above)
Comment (madlabman) [reply src/utils/blockstamp.py:34]: 953 `"` and 651 `'`, so no everywhere. And I've chosen the one used by black by default. AFAIR there was an intention to format the "src" someday? cc @F4ever you word, ser?

File: src/providers/execution/contracts/CSModule.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        """Returns node operators assumed to be stuck for the given frame (defined by the blockstamps)"""
+
+        l_block_number = int(self.w3.eth.get_block(l_block)["number"])  # type: ignore
+
+        by_no_id: Callable[[EventData], int] = lambda e: int(
+            e["args"]["nodeOperatorId"]
+        )  # TODO: Maybe it's int already?
+
+        events = sorted(self.get_stuck_keys_events(r_block), key=by_no_id)
+        for op_id, group in groupby(events, key=by_no_id):
```
[the remark this answers is not in the document]
Comment (madlabman) [reply src/providers/execution/contracts/CSModule.py:26]: I don't like this naming to be honest, but it's common to the codebase, you're right.

File: src/main.py [hunk 5]
```diff
@@ -78,6 +79,7 @@ def main(module_name: OracleModule):
         'lido_contracts': LidoContracts,
         'lido_validators': LidoValidatorsProvider,
         'transaction': TransactionUtils,
+        'csm': CSM,
```
[the remark this answers is not in the document]
Comment (madlabman) [reply src/main.py:82]: I thought about the same thing, but I think it's better to keep is as is for simplicity. Another idea I have is to you a lazy object wrapper around the class to instantiate the module if it's explicitly requested by the runtime code.

File: src/modules/submodules/oracle_module.py [hunk 6]
```diff
@@ -54,6 +54,7 @@ def run_as_daemon(self):
             logger.info({'msg': 'Startup new cycle.'})
             self.cycle_handler()
 
+    # TODO: We need a configurable timeout for CSM, just use a variable?
```
[the remark this answers is not in the document]
Comment (madlabman) [reply src/modules/submodules/oracle_module.py]: For CSM it doesn't make sense to set the timeout, just not to forget. Removed.

**Review by vgorkavenko — APPROVED** [review]

File: src/utils/blockstamp.py (hunk 2 above)
Comment (F4ever) [reply src/utils/blockstamp.py:34]: I try to use everywhere `'`.
And yeah, there is a setting in black to format to `"`, but no setting to format everything to `'`

**Pull request merged**